Documentation on reducing package size and custom CFLAGS

peterroelants opened the pull request:
## Change Summary

- Documentation update on how to custom compile *pydantic* when using pip install.
- Small change in `setup.py` to allow for custom CFLAGS when compiling.

### Setup.py CFLAGS
pydantic's [`setup.py` overwrites all custom CFLAGS](https://github.com/samuelcolvin/pydantic/blob/619ff261c9bdc2a612ef2fcfa17a005dfefbcf21/setup.py#L85) that a user might pass in when rebuilding pydantic. 
This change appends custom CFLAGS to the current CFLAGS used when building. There should not be any change in the current default behaviour.

By appending custom CFLAGS to the end it is possible to overwrite the default optimization flags currently used in pydantic. From the [gcc docs](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#:~:text=If%20you%20use%20multiple%20-O%20options):
> _If you use multiple -O options, with or without level numbers, the last such option is the one that is effective._

#### Performance impact of setting different CFLAGS

|      | size| benchmark avg time |
| --- | --- | --- |
| default `pip install` (from PyPI) | 45M | 70.7μs |
| `pip install --no-binary` (from PyPI) | 796K | 101.4μs |
| `pip install --no-binary` (from PyPI) with cython | 6.4M | 67.8μs |
| `pip install --no-binary` with cython CFLAG "-O3" (This branch) | 6.4M | 68.1μs |
| `pip install --no-binary` with cython CFLAG "-Os" (This branch) | 5.2M | 74.6μs |

## Related issue number

Issue https://github.com/samuelcolvin/pydantic/issues/2276 will be resolved by documenting how to reduce *pydantic* installation size. Additionally, users can pass in custom CFLAGs  to trade-off speed vs size.

## Checklist

* [x] Unit tests for the changes exist (Already existent)
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

peterroelants commented:
@PrettyWood I was wondering, should I request a review for this somewhere? And if so, where can I do this?
Or should I just leave it to be reviewed eventually?

PrettyWood commented:
I let Samuel merge PRs

samuelcolvin commented:
Hopefully I can make time to work on pydantic next week.

samuelcolvin commented on changes/2517-peterroelants.md:
```diff
@@ -0,0 +1,2 @@
+- Documentation update how to custom compile pydantic when using pip install.
+- Small change in setup.py to allow for custom CFLAGS when compiling.
```
```suggestion
Documentation update how to custom compile pydantic when using pip install, small change in `setup.py` to allow for custom CFLAGS when compiling.
```

I'm pretty sure adding multiple bullet points here would break (or confuse) [`make_history.py`](https://github.com/samuelcolvin/pydantic/blob/master/changes/make_history.py).

samuelcolvin commented on docs/install.md:
````diff
@@ -29,6 +32,25 @@ import pydantic
+### Performance vs package size trade-off
+
+Compiled binaries can increase the size of your Python environment. If for some reason you want to reduce size the size of your *pydantic* installation you can avoid installing any binaries using the [`pip --no-binary`](https://pip.pypa.io/en/stable/reference/pip_install/#install-no-binary) option. Make sure `cython` is not in your environment, or that you have the `SKIP_CYTHON` environment varia ...
+```
+SKIP_CYTHON=1 pip install \
+  --no-binary pydantic \
+  pydantic
+```
````
```bash
    SKIP_CYTHON=1 pip install --no-binary pydantic pydantic
    ```
    !!! note
        `pydantic` is repeated here intentionally - `--no-binary pydantic` tells `pip` you want no binaries for pydantic,
        the next `pydantic` tells `pip` which package to install.

I think this is clearer, people (including me) get confused about the usage of `--on-binary`.

(I can't make this a suggestion as it includes triple back quotes itself)

peterroelants replied:
Updated according to suggestion.

samuelcolvin commented on docs/install.md:
````diff
@@ -29,6 +32,25 @@ import pydantic
+```
+
+Alternatively you can re-compile *pydantic* with custom [build options](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html), e.g.:
+```
````
again, might be worth adding the language:

    ```bash

peterroelants replied:
Done

samuelcolvin commented on setup.py:
```diff
@@ -82,7 +82,9 @@ def extra(self):
-        os.environ['CFLAGS'] = '-O3'
+        # Set CFLAG to all optimizations (-O3)
+        # Any additional CFLAGS will be appended. Only the last optimization flag will have effect
+        os.environ['CFLAGS'] = '-O3 ' + os.environ.get('CFLAGS', '')
```
👍

samuelcolvin reviewed (commented):
otherwise LGTM.

samuelcolvin commented:
please update.

peterroelants commented:
> please update.

@samuelcolvin Thank you for the review. I updated this PR according to the suggestions, rebased the changes on the latest master, and pushed the changes here. This should be ready to merge.

PrettyWood commented on docs/install.md:
```diff
@@ -29,6 +32,27 @@ import pydantic
 
+### Performance vs package size trade-off
+
+Compiled binaries can increase the size of your Python environment. If for some reason you want to reduce size the size of your *pydantic* installation you can avoid installing any binaries using the [`pip --no-binary`](https://pip.pypa.io/en/stable/reference/pip_install/#install-no-binary) option. Make sure `Cython` is not in your environment, or that you have the `SKIP_CYTHON` environment varia ...
```
```suggestion
Compiled binaries can increase the size of your Python environment. If for some reason you want to reduce the size of your *pydantic* installation you can avoid installing any binaries using the [`pip --no-binary`](https://pip.pypa.io/en/stable/cli/pip_install/#install-no-binary) option. Make sure `Cython` is not in your environment, or that you have the `SKIP_CYTHON` environment variable set to avoid re-compiling *pydantic* libraries:
```

PrettyWood replied:
Typo + moved link

PrettyWood reviewed (changes requested):
And please rebase/merge to have updated deps to fix CI.
Otherwise looks good thanks!

peterroelants commented:
@PrettyWood I:
- Merged your suggestion,
- Squashed the commits into a single commit while rebasing on the latest master,
- Pushed changes to this branch.

Please review

PrettyWood commented:
https://smokeshow.helpmanual.io/4944455v3p5m500e732e/install/

![Screen Shot 2021-06-04 at 8 37 38 PM](https://user-images.githubusercontent.com/18406791/120848182-c52ef800-c574-11eb-898e-4ca13d9c0d07.png)

I think you need to remove the colon at the end of `!!! note:`
Please update

peterroelants commented:
> https://smokeshow.helpmanual.io/4944455v3p5m500e732e/install/
> 
> ![Screen Shot 2021-06-04 at 8 37 38 PM](https://user-images.githubusercontent.com/18406791/120848182-c52ef800-c574-11eb-898e-4ca13d9c0d07.png)
> 
> I think you need to remove the colon at the end of `!!! note:`
> Please update

resolved

please review
